Lms/split snapshot count timeframes

anathomical opened the pull request:
## WHAT
When fetching the current and previous counts for a given metric in Admin Snapshots, fetch the two values in parallel on the front-end instead of using a single endpoint/worker combo that we were using before which made two BigQuery queries in sequence.
## WHY
Parallelizing this data lookup should result in a reduction in the average time it takes before a user sees data on this page due to the cases where current data comes back alone in the new system faster than current+previous data could come back in the old system.
## HOW
- Add a new controller action for `previous_count` so that the front-end can fetch that data separately
- Create a new worker for processing `previous_count`, but it does exactly the same thing as `count` except it sends a different Pusher event name when it completes
- Modify the controller helpers so that when processing data for `previous_count` the previous period of the timeframe is passed to the worker
- Update count worker to only take a timeframe_start and timeframe_end for timeframe data rather than the old method of getting previous and current timeframe bounds (we now handle that split in the controller)
- Short-circuit the controller action so that if there is no valid timeframe data we don't bother trying to calculate anything (currently intended for use with `previous_count` when the timeframe is "all-time")
- Update the front-end to make two different API calls on component load (along with two different retry timers for those two loads)
- Register two different Pusher listeners so that we can handle callbacks for both current and previous count
- Move a bunch of the calculation logic that is intended to handle API data into an effect so that we can consolidate that code into one place despite having two different API calls that could come back in either order

### Notion Card Links
https://www.notion.so/quill/Split-Admin-Snapshot-Count-queries-into-two-different-API-Worker-Cache-flows-one-for-Current-Timefr-3796837d69c84170a4e110e3cfc48be9?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Refactor AdminSnapshot so that count and previous_count are separate

This will make it easier to request those two data points in parallel rather than serially from the front-end, which should lead to better performance

anathomical committed:
Update controller spec for tweaked signatures and names

anathomical committed:
Return early if timeframes are nil

This is only expected to happen in cases where the selected timeframe is "all-time" in which case there is no timeframe for previous_count

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/split-snapshot-count-timeframes

anathomical committed:
Update front-end to request current and previous period data separately

anathomical committed:
Add some additional params to component for testing

anathomical committed:
Update change class logic

This is really an update for testing.  In the old system, changeDirection was nulled by default and wasn't set during some tests.  In the new system, effects set the value of changeDirection all the time, so keeping consistent class behavior required a tweak

anathomical committed:
Update test snapshots with new parameters

anathomical committed:
Update method signature per linting rule Style/OptionalBooleanParameter

anathomical committed:
Tweak spec to align with new payload shape expectations

emilia-friedberg reviewed (approved):
LGTM. Thanks for your work on this.

eadams17 reviewed (approved):
LGTM!

dandrabik reviewed (changes requested):
Thanks for putting this together, glad we were able to get this in place quickly. The previous and current controller endpoints, workers, and requests are identical, so I think there's a way to DRY this up by having one endpoint, worker, and request and passing one additional param, e.g. `previous: true/false` or `current: true/false` or whatever param makes you feel is clearest. That would make long-term maintenance easier and prevent these paths from drifting. I think it'll be a pretty small refactor.

anathomical committed:
Updates based on PR feedback

Specifically, fold current and previous timeframe controller actions and workers into singles one with params to differentiate them

anathomical committed:
Add unused method param to allow for matching signatures

dandrabik commented on services/QuillLMS/app/controllers/snapshots_controller.rb:
```diff
@@ -134,10 +134,18 @@ def options
   private def retrieve_cache_or_enqueue_worker(worker)
 
-    previous_start, previous_end, current_start, current_end = Snapshots::Timeframes.calculate_timeframes(snapshot_params[:timeframe],
+    previous_start, previous_end, timeframe_start, timeframe_end = Snapshots::Timeframes.calculate_timeframes(snapshot_params[:timeframe],
```
Would this be cleaner if you passed `previous_timeframe` to the `calculate_timeframes` method and returned the correct start/end from there, versus swapping below. This current setup does 4 calculations for timeframes, then always throws away 2.

Looking at the calculate_timeframe code, you could use the `previous_timeframe` boolean to determine whether to run/return `[:previous_start, :previous_end]` or `[:current_start, :current_end] `

dandrabik reviewed (approved):
One small refactor suggestion, but otherwise approved. Thanks for pushing this along. Also, Let's make sure to do some testing on staging and prod to make sure this is working properly, in case the refactor caused any bugs.

anathomical committed:
A bit more updating based on code review

anathomical committed:
Fix a small bug in async direction calculation

anathomical committed:
Fix linting error

anathomical committed:
Fix a spec to work with the updated timeframe lib code

anathomical committed:
Fix signature mis-match in passing previous_timeframe param
